### ISSUE TITLE
max-jobs cannot exceed # of CPU threads

### ISSUE THREAD

**Issue by rickypai** [axboe/fio#1151]
**Please acknowledge the following before creating a ticket**

- [x] I have read the GitHub issues section of [REPORTING-BUGS](../blob/master/REPORTING-BUGS).

**Description of the bug:**

`--max-jobs` argument seems to not be respected when it's greater than the # of CPU threads.

**Environment**:

Using an [Intel i7-7920HQ](https://ark.intel.com/content/www/us/en/ark/products/97462/intel-core-i7-7920hq-processor-8m-cache-up-to-4-10-ghz.html) processor which has 4 cores and 8 threads.

```plaintext
$ nproc
8
```

```plaintext
$ system_profiler SPHardwareDataType SPSoftwareDataType -detailLevel mini
Hardware:

    Hardware Overview:

      Model Name: MacBook Pro
      Model Identifier: MacBookPro14,3
      Processor Name: Quad-Core Intel Core i7
      Processor Speed: 3.1 GHz
      Number of Processors: 1
      Total Number of Cores: 4
      L2 Cache (per Core): 256 KB
      L3 Cache: 8 MB
      Hyper-Threading Technology: Enabled
      Memory: 16 GB
      System Firmware Version: 429.50.1.0.0
      SMC Version (system): 2.45f4

Software:

    System Software Overview:

      System Version: macOS 11.0.1 (20B50)
      Kernel Version: Darwin 20.1.0
      Time since boot: 18 days 11:56
```

**fio version**:

```
$ fio --version
fio-3.19
```

**Reproduction steps**

```plaintext
$ fio --numjobs=16 --max-jobs=1 --name=test
test: (g=0): rw=read, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=psync, iodepth=1
error: maximum number of jobs (1) reached.
```

max number of jobs is 1 as defined by the argument.

```plaintext
$ fio --numjobs=16 --max-jobs=16 --name=test
test: (g=0): rw=read, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=psync, iodepth=1
error: maximum number of jobs (8) reached.
```

max number of jobs is still 8.

**Comment by sitsofe** [comment]
@rickypai hello,

Just to check - you're on Linux of some description and not macOS? Whichever case, could you update the Environment section of your report? Thanks!

**Comment by sitsofe** [comment]
Additionally:
- if you're on Linux are you running in a container?
- If you're on Linux what is the output of `lsipc` ?

**Comment by rickypai** [comment]
Sorry I'm on macOS. will update the environment section

**Comment by sitsofe** [comment]
@rickypai So the problem is that by default macOS has a VERY small IPC shm area (if you were on the latest fio you would get a different error message - https://github.com/axboe/fio/blob/2ee239fb084355e115cf8c2bf8051e8807c4222a/init.c#L495 ) which only has space for. eight jobs (so max-jobs doesn't relate to this low limit).

The key is ask macOS to change the amount of IPC memory available. I believe changing `kern.sysv.shmmax` and `kern.sysv.shmall` were enough to make larger job values work. Can you let us know what values work for you?

On another note, in order to help the next reporter how would you improve the bug template/help text to avoid missing information?

**Comment by rickypai** [comment]
> macOS IPC memory

Thanks a lot. I'll give that a try right now.

> latest fio

Sorry about that. I installed via homebrew which is usually good about keeping libraries up-to-date, so I neglected to check against github for the actual latest release.

> in order to help the next reporter how would you improve the bug template/help text to avoid missing information?

The environment context was extremely helpful in debugging this particular issue and potentially others. For this specifically, I'd probably update the "Environment" section in `REPORTING-BUGS` to: 1. remove note about build bugs. I originally omitted the OS info since I thought my issue wasn't a build one. 2. sample commands for determining OS info for those who are less familiar. (`lsipc` doesn't exist for mac, for example)

I might even go as far as adding more non-Linux specific info in the documentation. For example, the `max-jobs` documentation does mention shared-memory limits on Linux, but the lack of mention for other platform led me to think it's a Linux-specific issue. The recommended tweak `/proc/sys/kernel/shmmax` is also Linux-specific.

I can help with adding more documentation as I run into more of these cases. For context, I'm trying to use `fio` as a standardized benchmark across platforms.

Since we're on the topic of jobs, would it be correct to say that jobs on fio is akin to threads in other io benchmarking tools like CrystalDiskMark?

**Comment by sitsofe** [comment]
So, 1. being reworded makes sense to make it more OS inclusive (Linux is fio's original home and due to the nature of both, knowing the kernel version is useful on that particular platform). 2. I think we only need to start with the high level info (e.g. macOS 10.15) because different contexts require different extra info and we can drill down. Don't worry about the lsipc comment - I only asked because I wasn't 100% sure you weren't on Linux but I was hedging my bets if you check the wording :-).

> Since we're on the topic of jobs, would it be correct to say that jobs on fio is akin to threads in other io benchmarking tools like CrystalDiskMark?

Maybe :-) The answer depends on platform and options (see https://fio.readthedocs.io/en/latest/fio_doc.html#cmdoption-arg-thread ). I'll also caution (since you mentioned a Windows tool) there are [yet to be resolved reasons in fio](https://github.com/axboe/fio/issues/875) why a straight up comparison against CrystalDiskMark/[diskspd](https://github.com/microsoft/diskspd) won't give comparable numbers...

**Comment by sitsofe** [comment]
(Waiting on reporter to post results of tweaking macOS system parameters)

**Comment by sitsofe** [comment]
(Oh and if you're building recent fio by hand on macOS take a look at #1006 )

**Comment by rickypai** [comment]
Adjusting `shmmax` and `shmall` worked! Some questions:

1. For ioengine, It seems like most Linux tests use `ioengine=libaio`. Why is that even though `psync` is the default? Should I be using a different ioengine for macOS or leave it with the default `psync`?

2. I also ran into another issue with `max-jobs` that might not be OS-specific and more of my lack of understanding with fio.
   
   When running with `max-jobs` and multiple jobs defined in a config file, it seems like `max-jobs` needs to be equal to the sum of `numjobs` of all jobs configured even though I have `stonewall` configured:

```
[global]
direct=1
size=1g
iodepth=1
numjobs=1
loops=1
directory=tests
group_reporting
stonewall

[read-rand-4k-q32t8]
rw=randread
bs=4k
iodepth=32
numjobs=8
stonewall

[read-rand-4k-q32t1]
rw=randread
bs=4k
iodepth=32
numjobs=1
stonewall
```


```
fio CrystalDiskMark-read.conf --max-jobs=8 # fails
fio CrystalDiskMark-read.conf --max-jobs=9 # succeeds
```

I can open a new issue to keep the existing discussion with macOS limits separate.

Thank you so much!

**Comment by sitsofe** [comment]
(Unfortunately we're getting off topic for a bug report and we're into "[How do I/Why is?](https://github.com/axboe/fio/blob/master/.github/SUPPORT.md#general-questions)" questions. After this can you send future questions to the [fio mailing list](https://github.com/axboe/fio/blob/fio-3.23/README#L58) (note: the list **only** accepts plain text emails)?!)

> For ioengine, It seems like most Linux tests use ioengine=libaio. Why is that even though psync is the default?

Because[ `psync` is a synchronous ioengine so it can't submit at an iodepth greater than one](https://fio.readthedocs.io/en/latest/fio_doc.html#cmdoption-arg-iodepth). On Linux `libaio` used to be the best asynchronous ioengine (which is why you will see it in so many examples trying to push the maximum load on that platform). As it's Linux only, on macOS `libaio` is not available and your only fio option if you want to submit I/O asynchronously from one process/thread is [`posixaio`](https://fio.readthedocs.io/en/latest/fio_doc.html#cmdoption-arg-ioengine). Most programs are submitting I/O synchronously so `psync` is a reasonable default. I'll save the longer discussion topic about iodepth vs jobs for another time.

> I also ran into another issue with max-jobs [...]

We (upstream) don't support older versions of fio. [In new versions of fio](https://github.com/axboe/fio/releases) [max-jobs is actually ignored](https://github.com/axboe/fio/blob/fio-3.25/init.c#L2739-L2740) (also see the [commit that "removed" it](https://github.com/axboe/fio/commit/42eb99528738f7d613e345c2db2eae63555234f7)) so if you upgrade you won't see a difference any more ;-)

I'll close this issue as the original title topic has been resolved but feel free to submit documentation patches for the changes you mentioned in your [earlier comment](https://github.com/axboe/fio/issues/1151#issuecomment-749240299) and we'll try and review them. Thanks!

**Comment by rickypai** [comment]
thank you so much. I'll make use of the mailing list :)
